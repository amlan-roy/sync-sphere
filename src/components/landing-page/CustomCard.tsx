import React from "react";
import { Card, CardContent, CardFooter, CardHeader } from "../ui/card";
import { cn } from "@/lib/utils";

type CardProps = React.ComponentProps<typeof Card>;
type CustomCardProps = CardProps & {
  cardHeader?: React.ReactNode;
  cardContent?: React.ReactNode;
  cardFooter?: React.ReactNode;
};

const CustomCard: React.FC<CustomCardProps> = ({
  className,
  cardHeader,
  cardContent,
  cardFooter,
  ...props
}) => {
  return (
    <Card
      data-testid="custom-card__card"
      className={cn("w-[380px]", className)}
      {...props}
    >
      <CardHeader data-testid="custom-card__header">{cardHeader}</CardHeader>
      <CardContent
        className="grid
        gap-4
      "
        data-testid="custom-card__content"
      >
        {cardContent}
      </CardContent>
      <CardFooter data-testid="custom-card__footer">{cardFooter}</CardFooter>
    </Card>
  );
};

export default CustomCard;
