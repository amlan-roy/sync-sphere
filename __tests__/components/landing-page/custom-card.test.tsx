/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react";
import CustomCard from "@/components/landing-page/CustomCard";

describe("components/landing-page/CustomCard", () => {
  const defaultProps = {
    cardHeader: <div data-testid="card-test-header">Card Header</div>,
    cardContent: <div data-testid="card-test-content">Card Content</div>,
    cardFooter: <div data-testid="card-test-footer">Card Footer</div>,
  };
  it("renders correct header, content and footer", () => {
    const tree = render(<CustomCard {...defaultProps} />);

    const renderedCardHeader = tree.queryByTestId("card-test-header");
    const renderedCardContent = tree.queryByTestId("card-test-content");
    const renderedCardFooter = tree.queryByTestId("card-test-footer");

    expect(tree.getByTestId("custom-card__card")).toBeInTheDocument();
    expect(tree.getByTestId("custom-card__header").firstChild).toEqual(
      renderedCardHeader
    );
    expect(tree.getByTestId("custom-card__content").firstChild).toEqual(
      renderedCardContent
    );
    expect(tree.getByTestId("custom-card__footer").firstChild).toEqual(
      renderedCardFooter
    );
    expect(tree).toMatchSnapshot();
  });

  it("renders only header when only header is passed", () => {
    const tree = render(<CustomCard cardHeader={defaultProps.cardHeader} />);

    const renderedCardHeader = tree.queryByTestId("card-test-header");
    const renderedCardContent = tree.queryByTestId("card-test-content");
    const renderedCardFooter = tree.queryByTestId("card-test-footer");

    expect(tree.getByTestId("custom-card__card")).toBeInTheDocument();
    expect(tree.getByTestId("custom-card__header").firstChild).toEqual(
      renderedCardHeader
    );
    expect(renderedCardContent).toBe(null);
    expect(renderedCardFooter).toBe(null);
    expect(tree).toMatchSnapshot();
  });

  it("renders only content when only content is passed", () => {
    const tree = render(<CustomCard cardContent={defaultProps.cardContent} />);

    const renderedCardHeader = tree.queryByTestId("card-test-header");
    const renderedCardContent = tree.queryByTestId("card-test-content");
    const renderedCardFooter = tree.queryByTestId("card-test-footer");

    expect(tree.getByTestId("custom-card__card")).toBeInTheDocument();
    expect(tree.getByTestId("custom-card__content").firstChild).toEqual(
      renderedCardContent
    );
    expect(renderedCardHeader).toBe(null);
    expect(renderedCardFooter).toBe(null);
    expect(tree).toMatchSnapshot();
  });

  it("renders only footer when only footer is passed", () => {
    const tree = render(<CustomCard cardFooter={defaultProps.cardFooter} />);

    const renderedCardHeader = tree.queryByTestId("card-test-header");
    const renderedCardContent = tree.queryByTestId("card-test-content");
    const renderedCardFooter = tree.queryByTestId("card-test-footer");

    expect(tree.getByTestId("custom-card__card")).toBeInTheDocument();
    expect(tree.getByTestId("custom-card__footer").firstChild).toEqual(
      renderedCardFooter
    );
    expect(renderedCardHeader).toBe(null);
    expect(renderedCardContent).toBe(null);
    expect(tree).toMatchSnapshot();
  });
});
