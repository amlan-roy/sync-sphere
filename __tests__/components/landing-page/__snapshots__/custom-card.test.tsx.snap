// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/landing-page/CustomCard renders correct header, content and footer 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
        data-testid="custom-card__card"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
          data-testid="custom-card__header"
        >
          <div
            data-testid="card-test-header"
          >
            Card Header
          </div>
        </div>
        <div
          class="p-6 pt-0 grid gap-4"
          data-testid="custom-card__content"
        >
          <div
            data-testid="card-test-content"
          >
            Card Content
          </div>
        </div>
        <div
          class="flex items-center p-6 pt-0"
          data-testid="custom-card__footer"
        >
          <div
            data-testid="card-test-footer"
          >
            Card Footer
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
      data-testid="custom-card__card"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
        data-testid="custom-card__header"
      >
        <div
          data-testid="card-test-header"
        >
          Card Header
        </div>
      </div>
      <div
        class="p-6 pt-0 grid gap-4"
        data-testid="custom-card__content"
      >
        <div
          data-testid="card-test-content"
        >
          Card Content
        </div>
      </div>
      <div
        class="flex items-center p-6 pt-0"
        data-testid="custom-card__footer"
      >
        <div
          data-testid="card-test-footer"
        >
          Card Footer
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/landing-page/CustomCard renders only content when only content is passed 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
        data-testid="custom-card__card"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
          data-testid="custom-card__header"
        />
        <div
          class="p-6 pt-0 grid gap-4"
          data-testid="custom-card__content"
        >
          <div
            data-testid="card-test-content"
          >
            Card Content
          </div>
        </div>
        <div
          class="flex items-center p-6 pt-0"
          data-testid="custom-card__footer"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
      data-testid="custom-card__card"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
        data-testid="custom-card__header"
      />
      <div
        class="p-6 pt-0 grid gap-4"
        data-testid="custom-card__content"
      >
        <div
          data-testid="card-test-content"
        >
          Card Content
        </div>
      </div>
      <div
        class="flex items-center p-6 pt-0"
        data-testid="custom-card__footer"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/landing-page/CustomCard renders only footer when only footer is passed 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
        data-testid="custom-card__card"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
          data-testid="custom-card__header"
        />
        <div
          class="p-6 pt-0 grid gap-4"
          data-testid="custom-card__content"
        />
        <div
          class="flex items-center p-6 pt-0"
          data-testid="custom-card__footer"
        >
          <div
            data-testid="card-test-footer"
          >
            Card Footer
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
      data-testid="custom-card__card"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
        data-testid="custom-card__header"
      />
      <div
        class="p-6 pt-0 grid gap-4"
        data-testid="custom-card__content"
      />
      <div
        class="flex items-center p-6 pt-0"
        data-testid="custom-card__footer"
      >
        <div
          data-testid="card-test-footer"
        >
          Card Footer
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/landing-page/CustomCard renders only header when only header is passed 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
        data-testid="custom-card__card"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
          data-testid="custom-card__header"
        >
          <div
            data-testid="card-test-header"
          >
            Card Header
          </div>
        </div>
        <div
          class="p-6 pt-0 grid gap-4"
          data-testid="custom-card__content"
        />
        <div
          class="flex items-center p-6 pt-0"
          data-testid="custom-card__footer"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm w-[380px]"
      data-testid="custom-card__card"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
        data-testid="custom-card__header"
      >
        <div
          data-testid="card-test-header"
        >
          Card Header
        </div>
      </div>
      <div
        class="p-6 pt-0 grid gap-4"
        data-testid="custom-card__content"
      />
      <div
        class="flex items-center p-6 pt-0"
        data-testid="custom-card__footer"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
